src/ring_theory/nullstellensatz.lean

```diff
@@ -0,0 +1,184 @@
+  smul_mem' := λ p q hq x hx,
+    by simp only [hq x hx, algebra.id.smul_eq_mul, mul_zero, ring_hom.map_mul] }
+
+@[simp] lemma mem_vanishing_ideal_iff {V : set (σ → k)} {p : mv_polynomial σ k} :
+  p ∈ vanishing_ideal V ↔ ∀ x ∈ V, eval x p = 0 := ⟨λ hp x hx, hp x hx, λ hp x hx, hp x hx⟩
+
+lemma vanishing_ideal_anti_mono {A B : set (σ → k)} (h : A ≤ B) :
+  vanishing_ideal B ≤ vanishing_ideal A :=
+λ p hp x hx, hp x $ h hx
+
+lemma vanishing_ideal_bot : vanishing_ideal (∅ : set (σ → k)) = ⊤ :=
+le_antisymm le_top (λ p hp x hx, absurd hx (set.not_mem_empty x))
+
+lemma le_vanishing_ideal_zero_locus (I : ideal (mv_polynomial σ k)) :
+  I ≤ vanishing_ideal (zero_locus I) :=
+λ p hp x hx, hx p hp
+
+lemma zero_locus_vanishing_ideal_le (V : set (σ → k)) :
+  V ≤ zero_locus (vanishing_ideal V) :=
+λ V hV p hp, hp V hV
+
+theorem zero_locus_vanishing_ideal_galois_connection :
+  @galois_connection (ideal (mv_polynomial σ k)) (order_dual (set (σ → k))) _ _
+    zero_locus vanishing_ideal :=
+λ I V, ⟨λ h, le_trans (le_vanishing_ideal_zero_locus I) (vanishing_ideal_anti_mono h),
+  λ h, le_trans (zero_locus_anti_mono h) (zero_locus_vanishing_ideal_le V)⟩
+
+lemma mem_vanishing_ideal_singleton_iff (x : σ → k) (p : mv_polynomial σ k) :
+  p ∈ (vanishing_ideal {x} : ideal (mv_polynomial σ k)) ↔ (eval x p = 0) :=
+⟨λ h, h x rfl, λ hpx y hy, hy.symm ▸ hpx⟩
+
+lemma vanishing_ideal_singleton_is_maximal (x : σ → k) :
+  (vanishing_ideal {x} : ideal (mv_polynomial σ k)).is_maximal :=
+begin
+  have : (vanishing_ideal {x} : ideal (mv_polynomial σ k)).quotient ≃+* k := ring_equiv.of_bijective
+    (ideal.quotient.lift _ (eval x) (λ p h, (mem_vanishing_ideal_singleton_iff x p).mp h))
+    begin
+      refine ⟨_, λ z, ⟨(ideal.quotient.mk (vanishing_ideal {x})) (C z), by simp⟩⟩,
+      refine (ring_hom.injective_iff _).mpr (λ p hp, _),
+      obtain ⟨q, rfl⟩ := quotient.mk_surjective p,
+      rwa [quotient.lift_mk, ← mem_vanishing_ideal_singleton_iff, ← quotient.eq_zero_iff_mem] at hp,
+    end,
+  rw [← bot_quotient_is_maximal_iff, ring_equiv.bot_maximal_iff this],
+  exact bot_is_maximal,
+end
+
+instance vanishing_ideal_singleton_is_maximal' {x : σ → k} :
+  (vanishing_ideal {x} : ideal (mv_polynomial σ k)).is_maximal :=
+vanishing_ideal_singleton_is_maximal x
+
+lemma radical_le_vanishing_ideal_zero_locus (I : ideal (mv_polynomial σ k)) :
+  I.radical ≤ vanishing_ideal (zero_locus I) :=
+begin
+  intros p hp x hx,
+  rw ← mem_vanishing_ideal_singleton_iff,
+  rw radical_eq_Inf at hp,
+  refine (mem_Inf.mp hp) ⟨le_trans (le_vanishing_ideal_zero_locus I)
+    (vanishing_ideal_anti_mono (λ y hy, hy.symm ▸ hx)), is_maximal.is_prime' _⟩,
+end
+
+/-- Transfer a set of points into the language of the prime spectrum,
+  `vanishing_ideal_points_to_points` and `points_to_points_zero_locus` give the computation rules -/
+def points_to_points : set (σ → k) → set (prime_spectrum (mv_polynomial σ k)) :=
```

Comment:
Can't we give this as function on actual points, instead of sets of points?

Reply:
I made this change, and changed the corresponding lemmas as well. I have both directions for the corresponding `vanishing_ideal` lemma, but for the corresponding `zero_locus` lemma I only get have one direction (`point_to_point_zero_locus_le`), and I'm not sure if that the other direction holds in general?